taichi/runtime/gfx/runtime.cpp

```diff
@@ -476,13 +431,28 @@ void GfxRuntime::launch_kernel(KernelHandle handle, RuntimeContext *host_ctx) {
           }
         } else {
           ext_array_size[i] = host_ctx->array_runtime_sizes[i];
+          uint32_t access = uint32_t(
+              ti_kernel->ti_kernel_attribs().ctx_attribs.arr_access.at(i));
+
           // Alloc ext arr
           if (ext_array_size[i]) {
+            bool host_write =
+                access & uint32_t(irpass::ExternalPtrAccess::READ);
             auto allocated = device_->allocate_memory_unique(
-                {ext_array_size[i], /*host_write=*/true, /*host_read=*/true,
+                {ext_array_size[i], host_write, false,
                  /*export_sharing=*/false, AllocUsage::Storage});
             any_arrays[i] = *allocated.get();
             allocated_buffers.push_back(std::move(allocated));
+
+            bool host_read =
+                access & uint32_t(irpass::ExternalPtrAccess::WRITE);
+            if (host_read) {
+              auto allocated = device_->allocate_memory_unique(
+                  {ext_array_size[i], false, true,
+                   /*export_sharing=*/false, AllocUsage::Storage});
+              any_array_shadows[i] = *allocated.get();
+              allocated_buffers.push_back(std::move(allocated));
+            }
```

Comment:
So IIUC the existing problem of excessive data transfer delay is a result of massive memory access to host-side memory? And your solution is to force the kernels read from device-local memory and copy the results to output buffers in `any_arrays_shadows` after kernel executions?

Reply:
It seems that reading performance is fine and we don't currently have a host->device bottleneck. We do have a device->host bottleneck. When the memory is placed on the GPU, the host read is very slow. When the memory is placed on the CPU, GPU kernels become too slow. It seems that using buffer copy (and thus the underlying DMA block transfer hardware) is faster overall.